uniffi/src/lib.rs

```diff
@@ -258,110 +262,64 @@ fn run_bindings_generate_subcommand(default_lib_file: Option<std::ffi::OsString>
-                bindings::swift::write_swift_bindings(&ci, &out_dir)?;
-
-                println!("Compiling generated Swift bindings into module...");
-                bindings::swift::compile_swift_module(&ci, &out_dir)?;
-            }
-            _ => bail!(
-                "Somehow tried to generate bindings for unsupported language {}",
-                lang
-            ),
-        }
+        println!(
+            "Generating {} bindings into {}",
+            lang,
+            out_dir.to_str().unwrap_or("[UNPRINTABLE]")
+        );
+        bindings::write_bindings(&ci, &out_dir, lang.try_into()?)?;
     }
     println!("Done!");
     Ok(())
 }
 
 fn run_bindings_exec_subcommand(
-    use_target_dir: bool,
+    default_lib_file: Option<std::ffi::OsString>,
     command_args: &clap::ArgMatches,
 ) -> Result<()> {
-    let curdir = current_target_dir()?;
-    let target_dir = if use_target_dir {
-        Some(curdir.as_os_str())
-    } else {
-        None
-    };
     let script_file = command_args.value_of("script");
-    let lang = match command_args.value_of("language") {
-        Some(lang) => lang,
+    let lang: bindings::TargetLanguage = match command_args.value_of("language") {
+        Some(lang) => lang.try_into()?,
         None => {
             // Try to guess language based on script file extension.
-            if let None = script_file {
+            if script_file.is_none() {
                 bail!("No script file and no language specified, so I don't know what language shell to start")
             }
             let script_file_buf = PathBuf::from(script_file.unwrap());
             match script_file_buf.extension().unwrap_or_default().to_str() {
-                Some("kts") => "kotlin",
-                Some("py") => "python",
-                Some("swift") => "swift",
+                Some(ext) => ext.try_into()?,
                 _ => bail!("Cannot guess language of script file, please specify it explicitly"),
             }
         }
     };
-    match lang {
-        "kotlin" => {
-            bindings::kotlin::run_kotlin_script(target_dir, script_file)?;
-        }
-        "python" => {
-            bindings::python::run_python_script(target_dir, script_file)?;
-        }
-        "swift" => {
-            bindings::swift::run_swift_script(target_dir, script_file)?;
+    // If we have a default cdylib, make sure that it has the relevant bindings available
+    // and run from within its parent directory.
+    let target_dir = match default_lib_file {
+        None => None,
+        Some(lib_file) => {
+            let lib_file = PathBuf::from(lib_file);
+            match lib_file.parent() {
+                None => None,
+                Some(target_dir) => {
+                    let ci = bindings::get_component_interface_from_cdylib(&lib_file)?;
+                    bindings::write_bindings(&ci, target_dir, lang)?;
+                    bindings::compile_bindings(&ci, target_dir, lang)?;
```

Comment:
This now re-generates and re-compiles the bindings whenever you `cargo run -- exec` a script, but I think that's an improvement in ergonomics overall, as you don't have to remember to do that by hand.

Reply:
👍 When I was working on Swift, I’d forget to do `cargo run -- generate` before `exec` all the time. Definitely an improvement!